std/random.d

```diff
@@ -1336,6 +1330,271 @@ version (AnyARC4Random)
+        if (!fnCryptGenRandom) return -1;
+        if (fnCryptGenRandom(hProvider, len, cast(PBYTE) ptr))
+            return len;
+        else
+            return -1;
+    }
+}
+else version (Posix)
+{
+    version (linux)
+    {
+        version (X86) private enum NR_getrandom = 355;
+        else version (X86_64) private enum NR_getrandom = 318;
+        else version (ARM) private enum NR_getrandom = 384;
+        else version (AArch64) private enum NR_getrandom = 278;
+        else version (SPARC) private enum NR_getrandom = 347;
+        else version (SPARC64) private enum NR_getrandom = 347;
+        else version (Alpha) private enum NR_getrandom = 511;
+        else version (IA64) private enum NR_getrandom = 1339;
+        else version (PPC) private enum NR_getrandom = 359;
+        else version (PPC64) private enum NR_getrandom = 359;
+        else version (S390) private enum NR_getrandom = 349;
+        else version (SystemZ) private enum NR_getrandom = 349;
+        else version (HPPA) private enum NR_getrandom = 339;
+        else version (HPPA64) private enum NR_getrandom = 339;
+        else version (MIPS_O32) private enum NR_getrandom = 4353;
+        else version (MIPS_N32) private enum NR_getrandom = 6317;
+        else version (MIPS64) private enum NR_getrandom = 5313;
+        else version (SH)
+        {
+            static if (size_t.sizeof == 4)
+                private enum NR_getrandom = 373;
+            else static if (size_t.sizeof == 8)
+                private enum NR_getrandom = 384;
+        }
+
+        static if (is(typeof(NR_getrandom)))
+        {
+            // getrandom was introduced in Linux 3.17. If it's not available
+            // syscall will return -1 and errno will be set to ENOSYS.
+
+            /*
+             * Flags for getrandom(2)
+             *
+             * GRND_NONBLOCK    Don't block and return EAGAIN instead
+             * GRND_RANDOM      Use the /dev/random pool instead of /dev/urandom
+             */
+            private enum GRND_NONBLOCK = 0x0001;
+            private enum GRND_RANDOM = 0x0002;
+
+            /*
+                http://man7.org/linux/man-pages/man2/getrandom.2.html
+                If the urandom source has been initialized, reads of up to 256 bytes
+                will always return as many bytes as requested and will not be
+                interrupted by signals.  No such guarantees apply for larger buffer
+                sizes.
+            */
+            private extern(C) int syscall(size_t ident, size_t n, size_t arg1, size_t arg2) @nogc nothrow;
+        }
+    }
+
+    /+
+    Tries to use a system API to get entropy. Resources are not cached
+    due to the expectation that this function will not be called more
+    than once per program execution.
+
+    Params:
+        ptr = address at which to start writing
+        len = number of bytes to write
+
+    Returns:
+        Number of bytes successfully written up to `len`, or -1 on failure
```

Comment:
No, it's up to `len`:

> On success, getrandom() returns the number of bytes that were copied
       to the buffer buf.  This may be less than the number of bytes
       requested via buflen if either GRND_RANDOM was specified in flags and
       insufficient entropy was present in the random source or the system
       call was interrupted by a signal.

http://man7.org/linux/man-pages/man2/getrandom.2.html


Though for small `n` it's guaranteed to succeed:

> If the entropy pool has been initialized and the
       request size is small (buflen <= 256), then getrandom() will not fail
       with EINTR.  Instead, it will return all of the bytes that have been
       requested.

Reply:
I think we might as well make sure that we either get all the bytes we want, or that we get a failure.  That means that the function (which I'd still like to call `getSystemEntropy`) can have a common behaviour on all platforms where it exists at all.

It'd have to be `else if (errno != ENOSYS)`, wouldn't it ... ?

Reply:
BTW, a policy question: given the level this function operates at, might it not be better placed in `druntime` in the long run?  It feels like the sort of functionality that should live there rather than in phobos.

Reply:
> I think we might as well make sure that we either get all the bytes we want, or that we get a failure.

Oh, and note: given our existing use-cases, since we are never using `GRND_RANDOM` and we're always requesting < 256 bytes, we should never actually see those kinds of failure case.  So in the short term we might as well make it:

```D
if (result < len) return -1;
```

... with a `FIXME` or issue to improve things if we ever do want to start supporting larger quantites of entropy.

Reply:
> I think we might as well make sure that we either get all the bytes we want, or that we get a failure. 

Yep, my point!

> So in the short term we might as well make it:

Yeah, but then someone comes along and uses it for other bits in `std.random`, so I would feel safer if the "proper" checking is done on the beginning. Wrapping it in a loop isn't so hard after all.

> BTW, a policy question: given the level this function operates at, might it not be better placed in druntime in the long run? It feels like the sort of functionality that should live there rather than in phobos.

Well my PR to add just the numbers for syscalls is pending for four months:

https://github.com/dlang/druntime/pull/1984

So while I agree that this is low-level stuff, I don't think that we should block ourselves with that.
Also exposing something via druntime will mean that it's publicly visible and thus needs to have a well-though API, whereas this code can live in `private` land.

Reply:
> Wrapping it in a loop isn't so hard after all.

Yea, sounds good, then.  I wasn't sure if you were wanting that to actually be introduced now.

> So while I agree that this is low-level stuff, I don't think that we should block ourselves with that.

Oh, sure.  Although given that we can address the substantial issue (templatizing `unpredictableSeed`) without these entropy-generating functions, we could also just leave that to `druntime` and avoid the need for adding and then removing phobos code.  It would be nicer to do things patiently, but _right_, than to generate code churn through impatience.

That incidentally lends itself to better separation of concerns among all the changes.

> Also exposing something via druntime will mean that it's publicly visible and thus needs to have a well-though API

Er, are you suggesting that being `private` is an excuse for _not_ having a well-thought-through API? ;-)

Of course, I know what you _mean_, but that pressure to think things through properly is valuable.  I think we'd do better to implement things with that level of thought even if it's going to be private for now.  Apart from anything else, that makes it easier to upstream to druntime later.